subsys/net/ip/utils.c

```diff
@@ -518,3 +518,222 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
 
 	return false;
 }
+
+bool net_ipaddr_parse(const char *str, size_t str_len, struct sockaddr *addr)
+{
+	int count, i, ret;
+	int port;
+	char *ptr;
+
+	if (*str == '[') {
+#if defined(CONFIG_NET_IPV6)
+		/* IPv6 address with port number */
+		struct in6_addr *addr6;
+		char ipaddr[INET6_ADDRSTRLEN + 1];
+		char *endptr;
+		int end, len;
+
+		ptr = strstr(str, "]");
+		if (!ptr) {
+			return false;
+		}
+
+		len = min(INET6_ADDRSTRLEN, str_len);
+		end = min(len, ptr - (str + 1));
+
+		memcpy(ipaddr, str + 1, end);
+		ipaddr[end] = '\0';
+
+		addr6 = &net_sin6(addr)->sin6_addr;
+
+		ret = net_addr_pton(AF_INET6, ipaddr, addr6);
+		if (ret < 0) {
+			return false;
+		}
+
+		net_sin6(addr)->sin6_family = AF_INET6;
+
+		if ((ptr + 1) < (str + str_len) && *(ptr + 1) == ':') {
+			memcpy(ipaddr, ptr + 2, str_len - end);
+			ipaddr[str_len - end] = '\0';
+			port = strtol(ipaddr, &endptr, 10);
```

Comment:
Couldn't this port parsing be rewritten somehow to be common for both IPv6 & IPv4, instead of being duplicated?